### PR TITLE
Coerce request parameters

### PR DESCRIPTION
**Pull request by johnyherangi**
## Description

Automatically coerce request data to match handler function parameter types (primitive types only).

### Example

```ts
@Controller()
export class ReservationsController {
    @GET("/v1/reservations")
    public getReservations(
        @FromQuery("cancelled")
        cancelled: boolean,
        @FromQuery("page")
        page: number
    ) {
        console.log(cancelled) // e.g. true
        console.log(page) // e.g. 1
        // ...
    }
}
```

## Benefits

We can avoid code like this:
```ts
    public getReservations(@FromQuery("cancelled") cancelled: boolean) {
        const parsed = toBoolean(cancelled)
        // ...
    }

    public getReservations(@FromQuery("cancelled") cancelled: string | boolean) {
        cancelled = toBoolean(cancelled)
        // ...
    }
```

### CODE REVIEW

**Comment by johnmikk** [comment]
I like where this is going @johnyherangi, a few questions:
- Could it be somehow configurable to enable/disable the coercion behaviour?  There may be some situations where you don't want it
- How can people extend the coercion for other types that aren't included in your bundled `getCoercionFn`?

**Comment by johnyherangi** [comment]
> * Could it be somehow configurable to enable/disable the coercion behaviour?  There may be some situations where you don't want it

The coercion function is determined by the parameter type you are decorating, so if you don't want any coercion, your function parameter can just be `string`. The string coercion function is `param => param`

> * How can people extend the coercion for other types that aren't included in your bundled `getCoercionFn`?

I originally had a `coerce` function on `RequestOptions` that allowed extensions but I removed it because I wasn't sure how useful it would actually be. `@FromBody()` already converts the request body to an object,  and the other request parameters (path, query, header) are usually going to be `string`, `number` or `boolean`.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "lambaa",
-    "version": "2.0.0",
+    "version": "2.1.0",
```
Comment (jackholmes1010) [inline package.json]: is this a breaking change? Maybe we need to increment major version?

File: tests/parsing.tests.ts [hunk 2]
```diff
@@ -89,44 +122,68 @@ describe("request parsing tests", () => {
...
     })
 
     it("extracts query string parameter from request", async () => {
         const event = createAPIGatewayEvent({
             resource: "from_query_test",
             method: "GET",
             queryStringParameters: {
-                test: "test_query_param",
+                string: "string",
+                number: "5",
```
Comment (jackholmes1010) [inline tests/parsing.tests.ts:219]: what happens if this isn't actually a number? Controller gets `NaN`??

Wonder if we should throw a `RequestError`?

Comment (johnyherangi) [reply tests/parsing.tests.ts:219]: Hmm, yeah might be better to throw a request error

File: package.json (hunk 1 above)
Comment (johnyherangi) [reply package.json]: The `NaN` issue might make this a breaking change if people are using `number` for their function parameters already. I can increment the major version to be safe.

**Review by jackholmes1010 — APPROVED** [review]
lgtm

**Review by johnmikk — APPROVED** [review]
I'm happy with this, it's been sitting out for a while, so we should get it merged 🚀

File: package-lock.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "lambaa",
-    "version": "3.2.1",
+    "version": "3.4.0",
```
Comment (jackholmes1010) [inline package-lock.json]: can you update the version in `package.json`?

File: src/decorators/FromPath.ts [hunk 4]
```diff
@@ -1,9 +1,14 @@
...
 import { FROM_PATH_METADATA_KEY } from "../constants"
+import { RequestOptions } from "../types"
 
 /**
  * Extract a parameter from the request resource path.
  */
-export default function FromPath(name: string): ParameterDecorator {
+export default function FromPath(
+    name: string,
+    options: Omit<RequestOptions, "required"> = {}
```
Comment (jackholmes1010) [inline src/decorators/FromPath.ts]: can we remove this?

File: src/coerce.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import RequestError from "./RequestError"
+
+const getCoercionFn = (
+    target: any,
+    propertyKey: string | symbol,
+    index: number
+): ((param: string) => unknown) => {
+    const paramType = Reflect.getMetadata(
```
Comment (jackholmes1010) [inline src/coerce.ts:8]: can we make this optional access to be safe:

```typescript
const paramType = Reflect.getMetadata(
    "design:paramtypes",
    target,
    propertyKey
)?.[index]?.name
```

File: src/decorators/FromPath.ts (hunk 4 above)
Comment (johnyherangi) [reply src/decorators/FromPath.ts]: remove `options` entirely?

**Pull request merged**